pkg/pillar/agentbase/agent.go

```diff
@@ -13,51 +15,125 @@ import (
+		WarningTime: DefaultWarningTime,
+		ErrorTime:   DefaultErrorTime,
+		AgentOptions: AgentOptions{
+			NeedWatchdog:          true,
+			CheckAndCreatePidFile: true,
+			StartAgentLog:         true,
+			// Temporary default until all agents use this
+			SubscribeToGlobalConfig: false,
+			GlobalConfigSubscriptionOptions: pubsub.SubscriptionOptions{
+				AgentName:   "",
+				TopicImpl:   types.ConfigItemValueMap{},
+				Activate:    false,
+				WarningTime: DefaultWarningTime,
+				ErrorTime:   DefaultErrorTime,
+			},
+		},
+		GlobalConfig: types.DefaultConfigItemValueMap(),
+	}
 }
 
 // CliParams - stores all the common cli params
 type CliParams struct {
 	DebugOverride bool
+	Debug         bool
 }
 
+// AgentOptions - stores various options an agent may use to control different functions in Run()
+type AgentOptions struct {
+	NeedWatchdog                    bool
+	CheckAndCreatePidFile           bool
+	AddAgentCLIFlagsFnPtr           ContextCallbackFnType
+	ProcessAgentCLIFlagsFnPtr       ContextCallbackFnType
+	StartAgentLog                   bool
+	SubscribeToGlobalConfig         bool
+	GlobalConfigSubscriptionOptions pubsub.SubscriptionOptions
+}
+
+// ContextCallbackFnType - Defines a structure for the type of function we use to add and process cli flags
+type ContextCallbackFnType func()
+
+// GlobalConfigCallbackFn - Defines a structure for the type of function we use to modify global config items
+type GlobalConfigCallbackFn func(gcp *types.ConfigItemValueMap)
+
 // Context - a struct that represents a general agent context
 type Context struct {
 	CLIParams    CliParams
 	ErrorTime    time.Duration
 	WarningTime  time.Duration
 	AgentName    string
-	NeedWatchdog bool
+	AgentOptions AgentOptions
+
+	StillRunning *time.Ticker
+
+	SubGlobalConfig          pubsub.Subscription
+	GlobalConfig             *types.ConfigItemValueMap
+	GCInitialized            bool
+	GlobalConfigModifyChecks GlobalConfigCallbackFn
 }
 
 // processCLIFlags - Add flags common to all agents
 func processCLIFlags(agentBase AgentBase) {
-	debugPtr := flag.Bool("d", false, "Debug flag")
-	agentBase.AddAgentSpecificCLIFlags()
-	flag.Parse()
 	ctx := agentBase.AgentBaseContext()
-	ctx.CLIParams.DebugOverride = *debugPtr
-	agentBase.ProcessAgentSpecificCLIFlags()
+	flag.BoolVar(&ctx.CLIParams.Debug, "d", false, "Debug flag")
+	if ctx.AgentOptions.AddAgentCLIFlagsFnPtr != nil {
+		ctx.AgentOptions.AddAgentCLIFlagsFnPtr()
+	}
+	flag.Parse()
+	if ctx.AgentOptions.ProcessAgentCLIFlagsFnPtr != nil {
+		ctx.AgentOptions.ProcessAgentCLIFlagsFnPtr()
+	}
+	ctx.CLIParams.DebugOverride = ctx.CLIParams.Debug
 }
 
 // Run - a general run function that will handle all of the common code for agents
-func Run(agentSpecificContext AgentBase) {
-	processCLIFlags(agentSpecificContext)
-	ctx := agentSpecificContext.AgentBaseContext()
-	debugOverride := ctx.CLIParams.DebugOverride
-	if debugOverride {
+func Run(ps *pubsub.PubSub, ctx interface{}) {
+	agentBase := ctx.(AgentBase)
```

Comment:
Why not have the arg be of type AgentBase?

Reply:
I changed to interface because the global config needs the agent specific context. If I just passed in the agentbase the context for the global config would only be whatever the agentbase was. Should I make it so you can pass in the agent specific context through maybe the AgentOptions or should I leave it? Thanks.